http/middleware/log_request.go

```diff
@@ -2,44 +2,167 @@ package middleware
 		})
 	}
 }
+
+// A LogRequestRecord represents the fields that a LogRequest
+type LogRequestRecord struct {
+	BodySize       int    `json:"bodySize"`
+	Duration       int64  `json:"duration"`
+	Host           string `json:"host"`
+	ID             string `json:"id"`
+	IPAddr         string `json:"remoteAddr"`
+	Method         string `json:"method"`
+	Path           string `json:"path"`
+	Protocol       string `json:"protocol"`
+	Referrer       string `json:"referrer"`
+	ReqContentLen  int    `json:"contentLength"`
+	ReqContentType string `json:"contentType"`
+	Scheme         string `json:"scheme"`
+	Status         int    `json:"status"`
+	URI            string `json:"uri"`
+	UserAgent      string `json:"userAgent"`
+}
+
+// newRecord constructs a record from the values availabe in w & r.
+func newRecord(w *requestLogger, r *http.Request) LogRequestRecord {
+	// TODO(dlk): if there's a compelling reason for constructing a LogRequestRecord
+	// outside this package, this constructor and LogRequestRecord.attrs could be exported.
+	uri := new(url.URL)
+	*uri = *r.URL
+	q := r.URL.Query()
+	mask(q, passwordParam)
+	uri.RawQuery = q.Encode()
+
+	contLen, _ := strconv.Atoi(r.Header.Get(contentLenHeader))
+	id, _ := r.Context().Value(trails.RequestIDKey).(string)
+	ip, _ := r.Context().Value(trails.IpAddrKey).(string)
+
+	return LogRequestRecord{
+		BodySize:       w.bodySize,
+		Host:           r.Host,
+		ID:             id,
+		IPAddr:         ip,
+		Method:         r.Method,
+		Path:           r.URL.Path,
+		Protocol:       r.Proto,
+		Referrer:       r.Header.Get(referrerHeader),
+		ReqContentLen:  contLen,
+		ReqContentType: r.Header.Get(contentTypeHeader),
+		Scheme:         r.URL.Scheme,
+		Status:         w.status,
+		URI:            uri.RequestURI(),
+		UserAgent:      r.Header.Get(userAgentHeader),
+	}
+}
+
+func (r LogRequestRecord) attrs() []slog.Attr {
+	return []slog.Attr{
+		slog.Int64("duration", r.Duration),
+		slog.String("host", r.Host),
+		slog.String("id", r.ID),
+		slog.String("remoteAddr", r.IPAddr),
+		slog.String("method", r.Method),
+		slog.String("path", r.Path),
+		slog.String("protocol", r.Protocol),
+		slog.String("referrer", r.Referrer),
+		slog.Int("contentLength", r.ReqContentLen),
+		slog.String("contentType", r.ReqContentType),
+		slog.Int("bodySize", r.BodySize),
+		slog.String("scheme", r.Scheme),
+		slog.Int("status", r.Status),
+		slog.String("uri", r.URI),
+		slog.String("userAgent", r.UserAgent),
+	}
+}
+
+type requestLogger struct {
+	http.ResponseWriter
+	status   int
+	bodySize int
+}
+
+func (rl *requestLogger) Header() http.Header { return rl.ResponseWriter.Header() }
+
+// Unwrap exposes the underlying http.ResponseWriter.
+//
+// NOTE(dlk): include in order to simplify using [net/http.ResponseController].
+func (rl *requestLogger) Unwrap() http.ResponseWriter { return rl.ResponseWriter }
+func (rl *requestLogger) Write(b []byte) (int, error) {
+	size, err := rl.ResponseWriter.Write(b)
+	rl.bodySize += size
```

Comment:
is this costly performance wise, with the only purpose being to get response body size? Response body size in logs is something that's rarely useful so I can live without it if this is a perf hit.

Reply:
I don't have any benchmarking data to work with to make an assessment, but I can't think of a handler calling `Write` more than once (trails doesn't fwiw), which is what I naively conjecture is the way there'd be a meaningful perf impact. The most likely situation where that'd happen in our code base is if someone hands the `http.ResponseWriter` to a proc for writing directly to it. There's 1 case of that in `college-try`, not seeing one in `second-child` or `wall-to-wall`.